api/src/main/java/com/epam/pipeline/manager/datastorage/DataStorageManager.java

```diff
@@ -778,16 +779,19 @@ public List<DataStorageTagSearchResult> searchDataStorageItemByTag(
             final Map<Long, List<DataStorageTag>> results = new HashMap<>();
             final Map<Long, List<AbstractDataStorage>> storagesByRootId = dataStorageDao.loadDataStoragesByRootIds(
                     rootSearchResult.keySet()
-            ).stream().filter(
-                ds -> CollectionUtils.isEmpty(storageIdsToFilter) || storageIdsToFilter.contains(ds.getId())
-            ).collect(Collectors.groupingBy(AbstractDataStorage::getRootId));
+            ).stream()
+                    .filter(dataStorage -> permissionManager.storagePermission(dataStorage, READ_PERMISSION))
+                    .filter(
+                            ds -> CollectionUtils.isEmpty(storageIdsToFilter) || storageIdsToFilter.contains(ds.getId())
+                    ).collect(Collectors.groupingBy(AbstractDataStorage::getRootId));
 
             rootSearchResult.forEach((rootId, rootStorageTags) ->
                 rootStorageTags.forEach(dataStorageTag -> {
-                    final List<AbstractDataStorage> storages = storagesByRootId.get(rootId);
+                    final List<AbstractDataStorage> storages = storagesByRootId
+                            .getOrDefault(rootId, Collections.emptyList());
                     storages.stream().filter(
                         dataStorage -> dataStorageTag.getObject().getPath().startsWith(dataStorage.getPrefix())
-                    ).findFirst().ifPresent(dataStorage -> {
+                    ).max(Comparator.comparingInt(s -> s.getPrefix().length())).ifPresent(dataStorage -> {
```

Comment:
I think that the chosen mechanism of selecting only a single data storage per root is very fragile. We should either select a closest data storage on per tag basic or return tags associated with all storages.

Let's look at the following example

```
# we have two storages
/storage1
/storage1/storage2

# we have two objects
/storage1/file1
/storage1/storage2/file2

# current implementation will associate all found objects with storage2 even if they are actually part of storage1
/storage1/storage2file1
/storage1/storage2/file2
```

Reply:
Nah. What I've described is not entirely correct. The mechanism should work fine.

At the same time I would propose to keep only a single chain call per line otherwise it can be a little confusing :)

Reply:
In fact this works in other way:
In your example we will have results as:

/storage1/file1
/storage1/storage2/file2 

because on step
```
.filter(dataStorage -> dataStorageTag.getObject().getPath().startsWith(dataStorage.getPrefix()))
```
we will peek `storage1` for `file1` and `storage2` for `file2`